src/audio/selector/selector.c

```diff
@@ -83,6 +91,140 @@ static struct comp_dev *selector_new(const struct comp_driver *drv,
 	return dev;
 }
 
+static void set_selector_params(struct comp_dev *dev,
+				struct sof_ipc_stream_params *params)
+{
+}
+#else
+static void build_config(struct comp_data *cd)
+{
+	enum sof_ipc_frame valid_format;
+
+	cd->source_format = cd->md.base_cfg.audio_fmt.depth;
+	audio_stream_fmt_conversion(cd->md.base_cfg.audio_fmt.depth,
+				    cd->md.base_cfg.audio_fmt.valid_bit_depth,
+				    &cd->source_format,
+				    &valid_format,
+				    cd->md.base_cfg.audio_fmt.s_type);
+
+	audio_stream_fmt_conversion(cd->md.output_format.depth,
+				    cd->md.output_format.valid_bit_depth,
+				    &cd->sink_format,
+				    &valid_format,
+				    cd->md.output_format.s_type);
+
+	cd->config.in_channels_count = cd->md.base_cfg.audio_fmt.channels_count;
+	cd->config.out_channels_count = cd->md.output_format.channels_count;
+}
+
+static struct comp_dev *selector_new(const struct comp_driver *drv,
+				     struct comp_ipc_config *config,
+				     void *spec)
+{
+	struct micsel_data *ipc_process = spec;
+	struct comp_dev *dev;
+	struct comp_data *cd;
+	int ret;
+
+	comp_cl_info(&comp_selector, "selector_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+	dev->ipc_config = *config;
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	ret = memcpy_s(&cd->md, sizeof(cd->md), ipc_process, sizeof(*ipc_process));
+	assert(!ret);
+
+	build_config(cd);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void set_selector_params(struct comp_dev *dev,
+				struct sof_ipc_stream_params *params)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer __sparse_cache *source;
+	struct comp_buffer __sparse_cache *source_b;
+	struct list_item *sink_list;
+	int i;
+
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		params->channels = cd->config.in_channels_count;
+	else
+		params->channels = cd->config.out_channels_count;
+
+	params->rate = cd->md.base_cfg.audio_fmt.sampling_frequency;
+	params->frame_fmt = cd->source_format;
+
+	for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+		params->chmap[i] = (cd->md.output_format.ch_map >> i * 4) & 0xf;
+
+	/* update each sink format */
+	list_for_item(sink_list, &dev->bsink_list) {
+		struct comp_buffer *sink_b =
+			container_of(sink_list, struct comp_buffer, source_list);
+		struct comp_buffer __sparse_cache *sink = buffer_acquire(sink_b);
+		struct ipc4_audio_format *out_fmt = &cd->md.output_format;
```

Comment:
...and since you'll be doing one more revision of this PR, you can also move the `out_fmt` calculation out of the loop since it doesn't change. And actually if you define it at the function scope you can also use it in the above loop at line 171

Reply:
thanks